<?xml version="1.0"?>
<launch>

  <arg name='enable_v2v' default='true'/>
  <arg name='enable_detection' default='false'/>

  <!-- Ego vehicle -->
  <arg name='role_name_1' default='hero1'/>
  <arg name="spawn_point_1" default="-75, 174, 0, 0, 0, 90"/><!-- use comma separated format "x,y,z,roll,pitch,yaw" -->
  <arg name="vehicle_model_1" default='vehicle.toyota.prius'/>

  <arg name='role_name_2' default='hero2'/>
  <arg name="spawn_point_2" default="-70, 184.0, 0, 0, 0, 0"/><!-- use comma separated format "x,y,z,roll,pitch,yaw" -->
  <arg name="vehicle_model_2" default='vehicle.tesla.model3'/>

  <arg name='map_frame' default='/map'/>

  <include file="$(find carla_ros_bridge)/launch/setup_bridge.launch">
  </include>

  <include file="$(find carla_ros_bridge)/launch/spawn_obstacles_1.launch">
  </include>

<!-- VEHICLE 1 -->
<group ns="$(arg role_name_1)">

  <include file="$(find carla_ego_vehicle)/launch/carla_example_ego_vehicle.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name="vehicle_filter" value='$(arg vehicle_model_1)'/>
    <arg name='role_name' value='$(arg role_name_1)'/>
    <arg name="spawn_point" value="$(arg spawn_point_1)"/>
  </include>

   <include file="$(find carla_manual_control)/launch/carla_manual_control.launch">
    <arg name='role_name' value='$(arg role_name_1)'/>
  </include>

  <node name="costmap_node_$(arg role_name_1)" pkg="costmap_2d" type="costmap_2d_node" >
    <rosparam file="$(find era_gazebo)/config/$(arg role_name_1)_costmap_2d.yaml" command="load" ns="costmap" />
    <remap from="costmap_node_$(arg role_name_1)/costmap/costmap" to="local_map"/>
  </node>

  <group if="$(arg enable_v2v)">
    <include file="$(find era_gazebo)/launch/v2v.launch">
      <arg name="robot_name" value="$(arg role_name_1)" />
     <arg name="local_ip" value="127.0.0.1" />
     <arg name="remote_ip" value="127.0.0.1" />
      <arg name="local_port" value="52001" />
      <arg name="remote_port" value="52002" />
      <arg name="robot_base_frame" value="$(arg role_name_1)" />
      <arg name="map_frame" value="$(arg map_frame)" />
    </include>
  </group>

  <group if="$(arg enable_detection)">
    <include file="$(find era_gazebo)/launch/object_detection.launch">
      <arg name="image_topic" value="/carla/$(arg role_name_1)/camera/rgb/front/image_color"/>
      <arg name="detection_model" value="$(arg detection_model)" />
      <arg name="use_gpu" value="true" />
    </include>
  </group>

</group>
 
<!-- VEHICLE 2 -->

<group ns="$(arg role_name_2)">

  <include file="$(find carla_ego_vehicle)/launch/carla_example_ego_vehicle.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name="vehicle_filter" value='$(arg vehicle_model_2)'/>
    <arg name='role_name' value='$(arg role_name_2)'/>
    <arg name="spawn_point" value="$(arg spawn_point_2)"/>
  </include>

   <include file="$(find carla_manual_control)/launch/carla_manual_control.launch">
    <arg name='role_name' value='$(arg role_name_2)'/>
  </include>

  <node name="costmap_node_$(arg role_name_2)" pkg="costmap_2d" type="costmap_2d_node" >
    <rosparam file="$(find era_gazebo)/config/$(arg role_name_2)_costmap_2d.yaml" command="load" ns="costmap" />
    <remap from="costmap_node_$(arg role_name_2)/costmap/costmap" to="local_map"/>
  </node>

  <group if="$(arg enable_v2v)">
    <include file="$(find era_gazebo)/launch/v2v.launch">
      <arg name="robot_name" value="$(arg role_name_2)" />
      <arg name="local_ip" value="127.0.0.1" />
      <arg name="remote_ip" value="127.0.0.1" />
      <arg name="local_port" value="52002" />
      <arg name="remote_port" value="52001" />
      <arg name="robot_base_frame" value="$(arg role_name_2)" />
      <arg name="map_frame" value="$(arg map_frame)" />
    </include>
  </group>

  <group if="$(arg enable_detection)">
    <include file="$(find era_gazebo)/launch/object_detection.launch">
      <arg name="image_topic" value="/carla/$(arg role_name_2)/camera/rgb/front/image_color"/>
      <arg name="detection_model" value="$(arg detection_model)" />
    </include>
  </group>


</group> 

</launch>
